packages/squiggle-lang/src/rescript/Distributions/PointSetDist/PointSetDist_Scoring.res

```diff
@@ -1,46 +1,185 @@
+  let quot = esti /. answ
+  quot < 0.0 ? Error(Operation.ComplexNumberError) : Ok(-.answ *. logFn(quot))
+}
+
+module WithDistAnswer = {
+  // The Kullback-Leibler divergence
+  let integrand = (estimateElement: float, answerElement: float): result<
     float,
     Operation.Error.t,
   > =>
     // We decided that negative infinity, not an error at answerElement = 0.0, is a desirable value.
     if answerElement == 0.0 {
       Ok(0.0)
-    } else if predictionElement == 0.0 {
+    } else if estimateElement == 0.0 {
       Ok(infinity)
     } else {
-      let quot = predictionElement /. answerElement
-      quot < 0.0 ? Error(Operation.ComplexNumberError) : Ok(-.answerElement *. logFn(quot))
+      minusScaledLogOfQuot(~esti=estimateElement, ~answ=answerElement)
+    }
+
+  let rec sum = (~estimate: t, ~answer: t, ~combineFn, ~integrateFn, ~toMixedFn): result<
+    float,
+    Operation.Error.t,
+  > =>
+    switch (estimate, answer) {
+    | (Continuous(_), Continuous(_))
+    | (Discrete(_), Discrete(_)) =>
+      combineFn(integrand, estimate, answer)->E.R2.fmap(integrateFn)
+    | (_, _) =>
+      let esti = estimate->toMixedFn
+      let answ = answer->toMixedFn
+      switch (
+        Mixed.T.toContinuous(esti),
+        Mixed.T.toDiscrete(esti),
+        Mixed.T.toContinuous(answ),
+        Mixed.T.toDiscrete(answ),
+      ) {
+      | (
+          Some(estiContinuousPart),
+          Some(estiDiscretePart),
+          Some(answContinuousPart),
+          Some(answDiscretePart),
+        ) =>
+        E.R.merge(
+          sum(
+            ~estimate=Discrete(estiDiscretePart),
+            ~answer=Discrete(answDiscretePart),
+            ~combineFn,
+            ~integrateFn,
+            ~toMixedFn,
+          ),
+          sum(
+            ~estimate=Continuous(estiContinuousPart),
+            ~answer=Continuous(answContinuousPart),
+            ~combineFn,
+            ~integrateFn,
+            ~toMixedFn,
+          ),
+        )->E.R2.fmap(((discretePart, continuousPart)) => discretePart +. continuousPart)
+      | (_, _, _, _) => `unreachable state`->Operation.Other->Error
+      }
     }
+
+  let sumWithPrior = (
+    ~estimate: t,
+    ~answer: t,
+    ~prior: t,
+    ~combineFn,
+    ~integrateFn,
+    ~toMixedFn,
+  ): result<float, Operation.Error.t> => {
+    let kl1 = sum(~estimate, ~answer, ~combineFn, ~integrateFn, ~toMixedFn)
+    let kl2 = sum(~estimate=prior, ~answer, ~combineFn, ~integrateFn, ~toMixedFn)
+    E.R.merge(kl1, kl2)->E.R2.fmap(((kl1', kl2')) => kl1' -. kl2')
+  }
 }
 
-module LogScoreWithPointResolution = {
-  let logFn = Js.Math.log
-  let score = (
-    ~priorPdf: option<float => float>,
-    ~predictionPdf: float => float,
-    ~answer: float,
+module WithScalarAnswer = {
+  let score' = (~estimatePdf: float => float, ~answer: float): result<float, Operation.Error.t> => {
+    let density = answer->estimatePdf
+    if density < 0.0 {
+      Operation.PdfInvalidError->Error
```

Comment:
It's weird you do this here, but not in ``WithDistAnswer``. There should be a standard solution here. 

If we can validate PointSetDist though, we shouldn't need to worry about this.

Reply:
Story filed https://github.com/quantified-uncertainty/squiggle/issues/721 I think we should resolve this thread for the purposes of the current PR.